Add the support of the NonSemanticDebugPrintf

jiaolu opened the pull request:
Add the support of glsl "debugPrintfExt" and hlsl "printf" as like c/c++ printf instruction.
the LowerDebugPrintf pass would output printf "out variables" to the driver supplied buffer.

amdvlk-admin commented:
### Test summary for commit 60634a82c9bacf2bb8107cb0744c2af9dffd95ef
<details><summary>CTS tests (Failed: 0/207358)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 36990/69126 (53.5%)</li>
<li>Failed: 0/69126 (0.0%)</li>
<li>Not Supported: 32136/69126 (46.5%)</li><li>Warnings: 0/69126 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 37210/69104 (53.8%)</li>
<li>Failed: 0/69104 (0.0%)</li>
<li>Not Supported: 31894/69104 (46.2%)</li><li>Warnings: 0/69104 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 34228/69128 (49.5%)</li>
<li>Failed: 0/69128 (0.0%)</li>
<li>Not Supported: 34900/69128 (50.5%)</li><li>Warnings: 0/69128 (0.0%)</li>
</ul></details>


</ul></details>

amdvlk-admin commented:
### Test summary for commit 920c7154521febcb97a17a074dd7dfb807ac6d89
<details><summary>CTS tests (Failed: 0/207358)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 36990/69126 (53.5%)</li>
<li>Failed: 0/69126 (0.0%)</li>
<li>Not Supported: 32136/69126 (46.5%)</li><li>Warnings: 0/69126 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 37210/69104 (53.8%)</li>
<li>Failed: 0/69104 (0.0%)</li>
<li>Not Supported: 31894/69104 (46.2%)</li><li>Warnings: 0/69104 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 34228/69128 (49.5%)</li>
<li>Failed: 0/69128 (0.0%)</li>
<li>Not Supported: 34900/69128 (50.5%)</li><li>Warnings: 0/69128 (0.0%)</li>
</ul></details>


</ul></details>

jiaolu commented:
ping to review.

Flakebi commented on lgc/patch/LowerDebugPrintf.cpp:
```diff
@@ -0,0 +1,239 @@
+  }
+
+  for (auto callInst : callees) {
+    callInst->dropAllReferences();
```
I would hope that the `dropAllReferences` is not needed here, because all uses have been replaced above.

Flakebi commented on lgc/patch/LowerDebugPrintf.cpp:
```diff
@@ -0,0 +1,239 @@
+                                                 iterator_range<User::op_iterator> vars, BuilderBase &builder) {
+
+  // Printf output variables in DWORDs
+  SmallVector<Value *, 4> printArgs;
+  // Records printf output variables are 64bit or not
+  SmallVector<bool, 4> bit64Vector;
```
Same here about the inline size of the SmallVector.

Flakebi commented on lgc/patch/LowerDebugPrintf.cpp:
```diff
@@ -0,0 +1,239 @@
+  Value *maxOffset = builder.getInt64(1 << 31);
+  entryOffset = builder.CreateBinaryIntrinsic(Intrinsic::umin, entryOffset, maxOffset);
+
+  // Buffer Header is {BufferOffset_Loword, BufferOffset_Hiword, rerv0, rerv1};
```
`rerv` is for reserved?

Flakebi commented on llpc/test/shaderdb/general/PipelineCs_DebugPrintf.pipe:
```diff
@@ -0,0 +1,31 @@
+layout(local_size_x = 1, local_size_y = 1, local_size_z = 1) in;
+void main()
+{
+    debugPrintfEXT("Output:%d\n", gl_GlobalInvocationID.x);
```
Can you also add a test that has two printfs with different argument types? :)

jiaolu commented on lgc/include/lgc/state/Defs.h:
```diff
@@ -115,6 +115,7 @@ const static char NggPrimShaderEntryPoint[] = "lgc.shader.PRIM.main";
 const static char EntryPointPrefix[] = "lgc.shader.";
 const static char CopyShaderEntryPoint[] = "lgc.shader.COPY.main";
 const static char NullFsEntryPoint[] = "lgc.shader.FS.null.main";
+const static char LowerDebugPrintf[] = "lgc.debug.printf";
```
i think using dialect would be substantial rewrite and this pr sits in internal repo for quite some time.

amdvlk-admin commented:
### Test summary for commit 5ff86c07fca02ede742578b206ed1052f71aaabf
<details><summary>CTS tests (Failed: 0/207368)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 36988/69136 (53.5%)</li>
<li>Failed: 0/69136 (0.0%)</li>
<li>Not Supported: 32148/69136 (46.5%)</li><li>Warnings: 0/69136 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 37206/69104 (53.8%)</li>
<li>Failed: 0/69104 (0.0%)</li>
<li>Not Supported: 31898/69104 (46.2%)</li><li>Warnings: 0/69104 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 34225/69128 (49.5%)</li>
<li>Failed: 0/69128 (0.0%)</li>
<li>Not Supported: 34903/69128 (50.5%)</li><li>Warnings: 0/69128 (0.0%)</li>
</ul></details>


</ul></details>

jasilvanus commented on lgc/patch/LowerDebugPrintf.cpp:
```diff
@@ -0,0 +1,256 @@
+
+  auto offsetTy = PointerType::get(builder.getInt64Ty(), ADDR_SPACE_BUFFER_FAT_POINTER);
+  Value *bufferPtr = builder.CreateBitCast(debugPrintfBuffer, offsetTy);
+  // uint64_t offset = AtomicAdd64(offsetPtr, entrySize);
+  // maxOffset = 1<<31;
+  // offset = offset < maxOffset ? offset : maxOffset;
```
These lines look like unintended remainders of earlier versions.
Can these be removed?

jiaolu replied:
These comment c++ code to explain following generated llvm code.

jasilvanus commented on lgc/patch/LowerDebugPrintf.cpp:
```diff
@@ -0,0 +1,256 @@
+
+  SmallVector<Value *> outputVals = {builder.getInt32(loEntryheader), builder.getInt32(hiEntryheader)};
+
+  // Prepare the dword sequence of printf output variables
```
Maybe call reserve before the loop?

jasilvanus commented on lgc/patch/LowerDebugPrintf.cpp:
```diff
@@ -0,0 +1,256 @@
+
+// =====================================================================================================================
+// Convert value to the DWords, also append to output vector
+// @val: input value
```
Nit: Space before colon

jasilvanus commented on llpc/translator/lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -4074,14 +4080,62 @@ template <> Value *SPIRVToLLVM::transValueWithOpcode<OpExtInst>(SPIRVValue *cons
+// Handle extended instruction Nonsemantics.debugPrintf
+//
+// @param bi : A SPIR-V debugPrintf instruction
+// @param spvValues: Instruction spirv parameters
+// @param func: Which function to generate code
+// @param bb: Which basicblock to generate code
```
Nite: Spaces before colons

jasilvanus commented on llpc/translator/lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -4430,6 +4484,18 @@ template <> Value *SPIRVToLLVM::transValueWithOpcode<OpDot>(SPIRVValue *const sp
 
+// =====================================================================================================================
+// Translate OpString
+// @param spvValues: Instruction spirv parameters
```
Nit: space before colon

jasilvanus commented on llpc/translator/lib/SPIRV/libSPIRV/SPIRVType.h:
```diff
@@ -506,9 +506,9 @@ class SPIRVTypeStruct : public SPIRVType {
   SPIRVTypeStruct(SPIRVModule *M, SPIRVId TheId, const std::vector<SPIRVType *> &TheMemberTypes,
                   const std::string &TheName)
       : SPIRVType(M, 2 + TheMemberTypes.size(), OpTypeStruct, TheId), Literal(false) {
-    MemberTypeIdVec.resize(TheMemberTypes.size());
```
Maybe the `resize` was intended as `reserve` (and should remain as such)?

jasilvanus replied:
Did you intentionally not use a `reserve` here?

jiaolu replied:
https://github.com/GPUOpen-Drivers/llpc/pull/2348/commits/6da28f77f32803623a0542dba7ab8ecae34ba469#diff-412ad8a64497d0741782996ddb081f18b2a192924dd66996c8ba2b9485446f7a

right now it is 
"
MemberTypeIdVec.reserve(TheMemberTypes.size());
"

not resize

jasilvanus replied:
Oh, my bad, I mis-interpreted the diff. Thanks!

jasilvanus reviewed (commented):
Thanks for addressing the comments.
Only one inline comment on a resolved (but not changed) issue remains.

jiaolu commented:
> Thanks for addressing the comments. Only one inline comment on a resolved (but not changed) issue remains.

can you check again, the latest code is 
"reserve" , not "resize"

jasilvanus commented:
> can you check again, the latest code is "reserve" , not "resize"

Yes, thanks!

jiaolu commented on lgc/include/lgc/patch/LowerDebugPrintf.h:
```diff
@@ -0,0 +1,47 @@
+  llvm::Value *createDebugPrintf(llvm::Value *debugPrintfBuffer, llvm::Value *formatStr,
+                                 llvm::iterator_range<llvm::User::op_iterator> vars, lgc::BuilderBase &builder);
+  void getDwordValues(llvm::Value *val, llvm::SmallVector<llvm::Value *> &output,
+                      llvm::SmallVector<bool> &output64Bits);
```
Argument output type is changed. 
The output64Bits type following Jannik suggestion is replaced with llvm::SmallBitVector

amdvlk-admin commented:
6da28f77f32803623a0542dba7ab8ecae34ba469 Jenkins build error.
/home/amdvlk/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/patch/Patch.cpp:76:10: fatal error: llvm/Transforms/IPO/PassManagerBuilder.h: No such file or directory
 #include  llvm/Transforms/IPO/PassManagerBuilder.h 
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
[2875/3047] Building CXX object compiler/llpc/llvm/tools/lgc/CMakeFiles/LLVMlgc.dir/patch/Gfx6Chip.cpp.o
[2876/3047] Building CXX object compiler/llpc/llvm/tools/lgc/CMakeFiles/LLVMlgc.dir/elfLinker/NullFragmentShader.cpp.o

jiaolu commented:
retest this please

amdvlk-admin commented:
### Test summary for commit 7e179625fbbaa40a7a6578e2aa4b82587703d869
<details><summary>CTS tests (Failed: 0/207358)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 36986/69126 (53.5%)</li>
<li>Failed: 0/69126 (0.0%)</li>
<li>Not Supported: 32140/69126 (46.5%)</li><li>Warnings: 0/69126 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 37206/69104 (53.8%)</li>
<li>Failed: 0/69104 (0.0%)</li>
<li>Not Supported: 31898/69104 (46.2%)</li><li>Warnings: 0/69104 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.6, Gfx10</summary><ul>

<li>Passed: 34225/69128 (49.5%)</li>
<li>Failed: 0/69128 (0.0%)</li>
<li>Not Supported: 34903/69128 (50.5%)</li><li>Warnings: 0/69128 (0.0%)</li>
</ul></details>


</ul></details>

jiaolu commented:
ping for review.

Flakebi commented on lgc/patch/LowerDebugPrintf.cpp:
```diff
@@ -0,0 +1,257 @@
+  // offset = offset < maxOffset ? offset : maxOffset;
+  Value *entryOffset = builder.CreateAtomicRMW(AtomicRMWInst::Add, bufferPtr, builder.getInt64(entrySize), MaybeAlign(),
+                                               AtomicOrdering::Monotonic, SyncScope::SingleThread);
+  Value *maxOffset = builder.getInt64(1 << 31);
```
`1 << 31` is taken as a signed int and is sign-extended to an uint64_t.
The CsLowerDebugPrintf.lgc test shows that it ends up as `-2147483648 = 0xffffffff80000000`, but it should be `0x80000000`.

Flakebi reviewed (commented):
Found one more thing.
Thanks for adding tests. It would have slipped through without them.

jiaolu commented:
retest this please

amdvlk-admin commented:
### Test summary for commit ac91f8d2792cc84d17ad67c43415d2f01e6cf592
<details><summary>CTS tests (Failed: 0/207358)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 36986/69126 (53.5%)</li>
<li>Failed: 0/69126 (0.0%)</li>
<li>Not Supported: 32140/69126 (46.5%)</li><li>Warnings: 0/69126 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 37206/69104 (53.8%)</li>
<li>Failed: 0/69104 (0.0%)</li>
<li>Not Supported: 31898/69104 (46.2%)</li><li>Warnings: 0/69104 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.6, Gfx10</summary><ul>

<li>Passed: 34225/69128 (49.5%)</li>
<li>Failed: 0/69128 (0.0%)</li>
<li>Not Supported: 34903/69128 (50.5%)</li><li>Warnings: 0/69128 (0.0%)</li>
</ul></details>


</ul></details>

amdvlk-admin commented:
### Test summary for commit ac91f8d2792cc84d17ad67c43415d2f01e6cf592
<details><summary>CTS tests (Failed: 0/207358)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 36986/69126 (53.5%)</li>
<li>Failed: 0/69126 (0.0%)</li>
<li>Not Supported: 32140/69126 (46.5%)</li><li>Warnings: 0/69126 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 37206/69104 (53.8%)</li>
<li>Failed: 0/69104 (0.0%)</li>
<li>Not Supported: 31898/69104 (46.2%)</li><li>Warnings: 0/69104 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.6, Gfx10</summary><ul>

<li>Passed: 34225/69128 (49.5%)</li>
<li>Failed: 0/69128 (0.0%)</li>
<li>Not Supported: 34903/69128 (50.5%)</li><li>Warnings: 0/69128 (0.0%)</li>
</ul></details>


</ul></details>

nhaehnle commented on lgc/builder/InOutBuilder.cpp:
```diff
@@ -1448,6 +1448,23 @@ Value *InOutBuilder::CreateTaskPayloadAtomicCompareSwap(AtomicOrdering ordering,
+// Create debug printf operation, and write to the output debug buffer
+// @args : Printf variable parameters
+// @instName : Instance Name
+llvm::Value *lgc::InOutBuilder::CreateDebugPrintf(llvm::ArrayRef<llvm::Value *> args, const llvm::Twine &instName) {
```
Please remove unnecessary namespace prefixes in .cpp files: you shouldn't need the `llvm::` and `lgc::` here.

nhaehnle commented on lgc/interface/lgc/Builder.h:
```diff
@@ -196,6 +197,11 @@ class Builder : public BuilderCommon {
   virtual void recordShaderModes(llvm::Module *module) {}
 
+  // Find the resource node for the given set,binding
+  virtual const ResourceNode *findResourceNode(ResourceNodeType nodeType, unsigned descSet, unsigned binding) const {
+    return nullptr;
+  }
```
Since BuilderReplayer no longer derives from Builder, I think you can make this a pure virtual method now (` = 0`)

jiaolu replied:
The empty func here is trying to avoid to implement function in the builderImpl and its subsequent classes like inoutBuilder.
Make it pure virtual would break the compiling here.

nhaehnle reviewed (commented):
Thank you! I did notice some minor nits inline, but apart from that it LGTM

amdvlk-admin commented:
### Test summary for commit 0658b08071f88a269c97c804c4969f8db9619d75
<details><summary>CTS tests (Failed: 0/207358)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 36986/69126 (53.5%)</li>
<li>Failed: 0/69126 (0.0%)</li>
<li>Not Supported: 32140/69126 (46.5%)</li><li>Warnings: 0/69126 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 37206/69104 (53.8%)</li>
<li>Failed: 0/69104 (0.0%)</li>
<li>Not Supported: 31898/69104 (46.2%)</li><li>Warnings: 0/69104 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.6, Gfx10</summary><ul>

<li>Passed: 34225/69128 (49.5%)</li>
<li>Failed: 0/69128 (0.0%)</li>
<li>Not Supported: 34903/69128 (50.5%)</li><li>Warnings: 0/69128 (0.0%)</li>
</ul></details>


</ul></details>

jiaolu commented:
ping for review.

ruiling commented on llpc/translator/lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -4074,14 +4080,62 @@ template <> Value *SPIRVToLLVM::transValueWithOpcode<OpExtInst>(SPIRVValue *cons
+// @param bb: Which basicblock to generate code
+Value *SPIRVToLLVM::transDebugPrintf(SPIRVInstruction *bi, const ArrayRef<SPIRVValue *> spvValues, Function *func,
+                                     BasicBlock *bb) {
+  if (!getBuilder()->findResourceNode(ResourceNodeType::DescriptorBuffer, Vkgc::InternalDescriptorSetId,
```
I think `DescBuilder` will return poison if the resource node was not found. Maybe you can just skip printf lowering in lgc if the printf buffer pointer is poison. Then we don't need to expose findResourceNode() to Builder anymore.

ruiling replied:
I see the internal discussion, the argument was `InternalDescriptorSetId` usage should only happen in frontend. But now that DescBuilder already has the ability to return poison descriptor if not found, it should be easy to skip printf lowering in lgc pass without using `InternalDescriptorSetId`. The translation happened here is really trivial. Anyway, I don't object to the change if others are ok with this.

jiaolu replied:
Originally i want to have a flag from driver to tell llpc to enable/disable debubPrintf. Here i am using the findResourceNode to replacing that flag.  it  should happens as early as possible in the lowering passes and debubPrintf is hence known enabled/disabled explicitly.

ruiling replied:
A flag sounds good to me if we want to skip them early.

ruiling commented on llpc/translator/lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -4074,14 +4080,62 @@ template <> Value *SPIRVToLLVM::transValueWithOpcode<OpExtInst>(SPIRVValue *cons
+    Type *bufType = transType(spvStructType);
+
+    m_debugOutputBuffer =
+        new GlobalVariable(*m_m, bufType, false, GlobalValue::ExternalLinkage, nullptr, "debugOutputBuffer", nullptr,
```
Can we use `getBuilder()->CreateGetDescPtr()` here to get the printf buffer descriptor? then you don't need to create the globalVariable and the metadata here.

jiaolu replied:
This was discussed in the internally before. Using global variables can entail cheap optimization. E.g.  for several raytracing shaders, and each shader has several debugPrintf.  Global variable can be merged into one after linking, and these numerous debug printf can share the same desc which is lowered in the beginning of the entry point.

ruiling replied:
Sounds reasonable. Could you please add the comment into the source file?

nhaehnle commented:
Can you please rebase, there are some conflicts?

amdvlk-admin commented:
72b41692c30f8eca7930d06521b5f390950c15ad Jenkins build error.
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/builder/BuilderRecorder.cpp:348:10: error: ‘m_pipelineState’ was not declared in this scope
   return m_pipelineState->findResourceNode(nodeType, descSet, binding).second;
          ^~~~~~~~~~~~~~~
[2844/3047] Building CXX object compiler/llpc/llvm/tools/lgc/CMakeFiles/LLVMlgc.dir/builder/BuilderReplayer.cpp.o
[2845/3047] Building CXX object compiler/llpc/llvm/tools/lgc/CMakeFiles/LLVMlgc.dir/builder/DescBuilder.cpp.o
[2846/3047] Building CXX object compiler/llpc/llvm/tools/lgc/CMakeFiles/LLVMlgc.dir/builder/ArithBuilder.cpp.o

amdvlk-admin commented:
### Test summary for commit b70f49507c85bc9d3788dc2f87cd100babc313e7
<details><summary>CTS tests (Failed: 0/208316)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 36991/69452 (53.3%)</li>
<li>Failed: 0/69452 (0.0%)</li>
<li>Not Supported: 32461/69452 (46.7%)</li><li>Warnings: 0/69452 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 37209/69420 (53.6%)</li>
<li>Failed: 0/69420 (0.0%)</li>
<li>Not Supported: 32211/69420 (46.4%)</li><li>Warnings: 0/69420 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.6, Gfx10</summary><ul>

<li>Passed: 34225/69444 (49.3%)</li>
<li>Failed: 0/69444 (0.0%)</li>
<li>Not Supported: 35219/69444 (50.7%)</li><li>Warnings: 0/69444 (0.0%)</li>
</ul></details>


</ul></details>

jiaolu commented:
> Can you please rebase, there are some conflicts?

Okay. the pr is rebased

jiaolu commented:
ping for review

jiaolu commented:
commits are rebased

amdvlk-admin commented:
### Test summary for commit 91379997a05797bbcd34a6f9ffb67bc1f737decc
<details><summary>CTS tests (Failed: 0/208297)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 36986/69439 (53.3%)</li>
<li>Failed: 0/69439 (0.0%)</li>
<li>Not Supported: 32453/69439 (46.7%)</li><li>Warnings: 0/69439 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 37206/69417 (53.6%)</li>
<li>Failed: 0/69417 (0.0%)</li>
<li>Not Supported: 32211/69417 (46.4%)</li><li>Warnings: 0/69417 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.6, Gfx10</summary><ul>

<li>Passed: 34225/69441 (49.3%)</li>
<li>Failed: 0/69441 (0.0%)</li>
<li>Not Supported: 35216/69441 (50.7%)</li><li>Warnings: 0/69441 (0.0%)</li>
</ul></details>


</ul></details>

nhaehnle commented on lgc/patch/LowerDebugPrintf.cpp:
```diff
@@ -0,0 +1,260 @@
+    callInst->eraseFromParent();
+  }
+
+  if (m_elfInfos.size() == 0)
+    return PreservedAnalyses::none();
```
Please just use `m_elfInfos.empty()`.

Also, the logic here still looks wrong to me:

* If m_elfInfos is empty, then this pass made no changes. So, we should return `PreservedAnalyses::all()`: all analyses were preserved.
* If m_elfInfos is not empty, then this pass did make changes. This means some analyses need to be discarded. This pass doesn't change the CFG, so it should return `PreservedAnalyses::allInSet<CFGAnalyses>()`.

nhaehnle reviewed (commented):
I think this is almost good to go, just one more inline comment.

amdvlk-admin commented:
### Test summary for commit 113738fb9ecae63b964e7c91944d5a218ffe850b
<details><summary>CTS tests (Failed: 0/208307)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 36988/69449 (53.3%)</li>
<li>Failed: 0/69449 (0.0%)</li>
<li>Not Supported: 32461/69449 (46.7%)</li><li>Warnings: 0/69449 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 37206/69417 (53.6%)</li>
<li>Failed: 0/69417 (0.0%)</li>
<li>Not Supported: 32211/69417 (46.4%)</li><li>Warnings: 0/69417 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.6, Gfx10</summary><ul>

<li>Passed: 34225/69441 (49.3%)</li>
<li>Failed: 0/69441 (0.0%)</li>
<li>Not Supported: 35216/69441 (50.7%)</li><li>Warnings: 0/69441 (0.0%)</li>
</ul></details>


</ul></details>

nhaehnle reviewed (approved):
Thanks, LGTM!

jiaolu commented on lgc/include/lgc/patch/LowerDebugPrintf.h:
```diff
@@ -0,0 +1,66 @@
+  void setupElfsPrintfStrings();
+  llvm::DenseMap<uint64_t, ElfInfo> m_elfInfos;
+  PipelineState *m_pipelineState = nullptr;
+  PipelineShadersResult *m_pipelineShaders = nullptr;
```
> These two fields are unused and give a warning on clang.

Thanks for head-up. created https://github.com/GPUOpen-Drivers/llpc/pull/2396 for this.
